[1.20.5] ModelDataManager cleanup

embeddedt opened the pull request:
This PR is largely based on design changes suggested by @PepperCode1 (from whom I would appreciate a review, if they have time/interest).

There are several goals that I aim to address. I list them below with the corresponding changes.

1. Remove the requirement for custom `BlockGetter`s to provide a full `ModelDataManager`. To do this, `getModelData(BlockPos)` was introduced on our BlockGetter extension and `getModelDataManager()` was moved down the inheritance chain to our Level extension.
2. Allow mods to snapshot the state of the model data manager for a given section without needing to rely on internal API. This was done by adding a `getAt(SectionPos)` method to the `ModelDataManager`. This API was designed to give mods the utmost flexibility. It returns an unmodifiable reference to the map for a given section. Mods may iterate over this and assemble a larger map (as NeoForge does in its own `Snapshot`) or clone the map as-is if they just want a snapshot for the single section.
3. With (1) and (2) implemented, the inheritance chain of `ModelDataManager` can be simplified. I removed the `Active` subclass and refactored all its contents into the parent class. The `Snapshot` static class remains, but it is marked as internal as we only use it in our patches to `RenderChunkRegion`. Mods shouldn't need to use it, as they can implement a snapshot tailored to their own purposes with the changes in (2).
4. To simplify the API surface, I made getters return `ModelData.EMPTY` when model data does not exist, rather than `null`. This means `getAtOrEmpty` is no longer required. If a mod *absolutely* needs to know that the empty model data was provided by a block entity as opposed to being the default return value, it can use `containsKey` on the map returned by (2). I don't see this as a very likely scenario, and it makes the API more ergonomic to remove the nullability.

(N.B.: The change to the `ChunkAccess` patch was done so that the model data manager could be accessed from `getWorldForge()` without requiring a cast to `Level`. The only override of this method returns a `Level`, anyways.)

I'm happy to adjust the design in response to feedback. Obviously, this PR is targeted at 1.20.5 due to the breaking changes.

embeddedt committed:
ModelDataManager cleanup + API surface reduction

PepperCode1 commented on src/main/java/net/neoforged/neoforge/common/extensions/IBlockGetterExtension.java:
```diff
@@ -59,11 +59,12 @@ default AuxiliaryLightManager getAuxLightManager(ChunkPos pos) {
-    default ModelDataManager getModelDataManager() {
-        return null;
+    default ModelData getModelData(BlockPos pos) {
+        return ModelData.EMPTY;
```
My proposal was based on Fabric API's design. There, the default implementation of this method works like `BlockEntity be = getBlockEntity(pos); return be != null : be.getModelData() : ModelData.EMPTY`. While this isn't an appropriate implementation for asynchronous scenarios and doesn't consider the current level's cache, it could still be useful as many modded scenarios I have seen are not asynchronous and are independent of the current level. Either default implementation is fine for my use cases but this is worth considering.

embeddedt replied:
This seems like a reasonable choice as well; I'll leave it open for others to comment as I personally am undecided which is better.

XFactHD replied:
I don't think that's a good idea. One of the most prevalent uses of custom `Block(AndTint)Getter`s is for tricks related to `IBakedModelExtension#getModelData()` and if a mod forgets to override the extension method in their custom implementation, then you've got an almost untraceable bug that reintroduces the exact concurrency issues the model data system aims to prevent. If it defaults to empty data though, it's almost immediately obvious what's wrong when a mod that heavily relies on model data interacts with one that uses a custom `Block(AndTint)Getter` (I had this case with FramedBlocks and Athena where the latter didn't forward the `ModelDataManager` from the wrapped `BlockGetter`).

Technici4n replied:
To add to what XFact said, there can also be mods that have a very expensive model data computation and rely on the cache in the level. It's not a good idea to bypass such a cache.

XFactHD commented on patches/net/minecraft/world/level/chunk/ChunkAccess.java.patch:
```diff
@@ -111,5 +111,5 @@
 +    }
 +    @Nullable
-+    public net.minecraft.world.level.LevelAccessor getWorldForge() { return null; }
++    public net.minecraft.world.level.Level getWorldForge() { return null; }
```
If the return type is changed anyway, then we can just rename this to `getLevel()` and remove the delegating override in `LevelChunk` since there is no obfuscation anymore that would cause the vanilla method in `LevelChunk` to not override this one in a production environment.

XFactHD commented on src/main/java/net/neoforged/neoforge/common/extensions/IBlockExtension.java:
```diff
@@ -852,7 +852,7 @@ default boolean canConnectRedstone(BlockState state, BlockGetter level, BlockPos
      * As such, if you need any data from your {@link BlockEntity}, you should put it in {@link ModelData} to guarantee
      * safe concurrent access to it on the client.<br/>
      * {@link ModelDataManager#getAt(BlockPos)} will return the {@link ModelData} for the queried block,
```
All references to `ModelDataManager` on this extension method and the appearance one should be replaced with references to the new model data accessor.

XFactHD reviewed (changes requested):
Looks good overall, just a few nitpicks.

PepperCode1 commented on src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java:
```diff
@@ -23,142 +22,130 @@
-     * Users should not be instantiating this themselves unless they know what they're doing.
+     * Snapshot the state of this manager for all sections in the volume specified by the given section coordinates.
+     *
+     * @throws IllegalArgumentException if this is a snapshot and the given region doesn't match the snapshot's region
```
This is no longer the case as `Snapshot` does not extend `ModelDataManager`.

PepperCode1 commented on src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java:
```diff
@@ -23,142 +22,130 @@
+     * @return an (unmodifiable) map containing the {@link ModelData} stored for the given chunk section
      */
-    public abstract ModelData getAtOrEmpty(BlockPos pos);
+    public Long2ObjectMap<ModelData> getAt(SectionPos pos) {
```
This should be annotated with `UnmodifiableView`.

PepperCode1 commented on src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java:
```diff
@@ -23,142 +22,130 @@
-        @Override
-        public ModelData getAtOrEmpty(BlockPos pos) {
-            return Objects.requireNonNullElse(getAt(pos), ModelData.EMPTY);
+    public ModelDataManager.Snapshot snapshotSectionRegion(int sectionMinX, int sectionMinY, int sectionMinZ, int sectionMaxX, int sectionMaxY, int sectionMaxZ) {
```
Since `Snapshot` is `Internal`, this method should be too; however, I'm not sure `Snapshot` should be internal. It's just a wrapper around a `Long2ObjectMap`, but so is the return of `Long2ObjectMaps.unmodifiable(map)`. I think the copying code in the constructor of `Snapshot` should be moved to this method, this method should return `@Unmodifiable Long2ObjectMap<ModelData>`, and `Snapshot` should be removed.

embeddedt replied:
Seems reasonable. We only ever use the `get` function in Neo itself, so I made it return a `Long2ObjectFunction` (this makes it trivial to implement the empty snapshot by just making a constant lambda, as opposed to needing to make a custom empty map with a non-null return value). Aside from that I followed your suggestion.

embeddedt committed:
Remove `ModelDataManager.Snapshot` and replace it with a Long2ObjectFunction

embeddedt committed:
Annotate ModelDataManager section getter as UnmodifiableView

embeddedt committed:
Replace some missed references to ModelDataManager#getAt

embeddedt committed:
Rename getWorldForge -> getLevel and remove the patched-in delegate

embeddedt committed:
Null-check the return value of BlockEntity#getModelData

Now that all the getters are expected to return EMPTY, we don't
want a null value sneaking through from a BE and not being attributed
to the correct mod

embeddedt committed:
Remove unneeded import

embeddedt committed:
Spotless

Technici4n commented on src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java:
```diff
@@ -6,211 +6,170 @@
-     * @param pos The position to query
+     * Provides all the model data for a given chunk section. This is useful for mods which wish to retrieve
+     * a fast view of the model data for a single section in a level.
+     * <p></p>
```
what the hell - stop doing that :P

Technici4n reviewed (commented):
Looks good

PepperCode1 commented on src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java:
```diff
@@ -91,14 +93,25 @@ public ModelData getAt(BlockPos pos) {
+     * The snapshot will return {@link ModelData#EMPTY} for nonexistent keys.
      */
-    public ModelDataManager.Snapshot snapshotSectionRegion(int sectionMinX, int sectionMinY, int sectionMinZ, int sectionMaxX, int sectionMaxY, int sectionMaxZ) {
+    @Unmodifiable
```
The `Unmodifiable` annotation only applies to collections and maps as per its documentation, so applying it to a function doesn't make sense.

PepperCode1 reviewed (changes requested):
All changes look good except the last nitpick I commented on.

embeddedt committed:
Fix HTML in javadoc

embeddedt committed:
Remove Unmodifiable from snapshotSectionRegion

Long2ObjectFunction is effectively immutable anyway if the provider
wants it to be

embeddedt commented:
Force pushed to rebase on 1.20.5-pre2, old commits are otherwise unchanged. (I applied review suggestions, of course.)

XFactHD reviewed (approved):
:shipit:

embeddedt commented:
Force pushed to rebase on 1.20.5-pre3, old commits are otherwise unchanged.
